Feature: UDP Implemented

coronas2k opened the pull request:
#99 #495 #566 #610

I've added the ability to check UDP services, just needed to tweak `fsockopen` since it assumes TCP by default, so essentially I've added a MySQL enum field which defaults to TCP, edited the templates, added the English localization. I was nearly going to implement auto selection by port via javascript but it would be inconvenient to the user if they wanted to override it wrestling with event listeners, so I assume TCP by default unless the user overrides it by selecting UDP since most people are monitoring TCP services.

Let me know if you need more tweaking or suggestions.

- Nayef

coronas2k committed:
protocol implementation
added protocol field to SQL & StatusUpdater.

coronas2k committed:
Feature(Protocol): UDP Implemented
- updated controllers
- templates to accommodate tcp or udp
- English localization label

coronas2k committed:
optimized: TCP Selected by default
- assume TCP by default was considering jquery autoselect but some ports can be UDP or TCP it would be inconvenient to the user

coronas2k commented:
What would be the best solution to sign the commits? Should I rebase? I wasn't aware you required signing.

coronas2k commented:
I rebased my fork to add my signature to the commits, just up to the commits I added.

coronas2k commented:
I'm able to implement auto selection by port via javascript initially and let the user change it later, let me know if that would be something you'd want.

TimZ99 commented on src/psm/Util/Server/Updater/StatusUpdater.php:
```diff
@@ -220,7 +220,7 @@ protected function updateService($max_runs, $run = 1)
         $starttime = microtime(true);
 
-        $fp = @fsockopen($this->server['ip'], $this->server['port'], $errno, $this->error, $timeout);
+        $fp = @fsockopen($this->server['protocol'] . '://' . $this->server['ip'], $this->server['port'], $errno, $this->error, $timeout);
```
Line to long, max 120 characters. 
@fsockopen(
$this->server['ip'],
$this->server['port'],
$errno,
$this->error,
$timeout
);

coronas2k replied:
Sure, can we create a CONTRIBUTING.md in the future to specify what standard we follow (eg: PSR1, PSR2, PSR12) & how to sign commits. 

Makes it clearer to submit pulls and easier to set up our phpcs.

TimZ99 commented on src/psm/Util/Install/Installer.php:
```diff
@@ -239,6 +239,7 @@ protected function installTables()
 				`server_id` int(11) unsigned NOT NULL AUTO_INCREMENT,
 				`ip` varchar(500) NOT NULL,
 				`port` int(5) NOT NULL,
+                `protocol` enum('tcp','udp') NOT NULL default 'tcp',
```
Upgrade is missing.

coronas2k replied:
I'm assuming the next release will be 3.5.0? I'm naming the upgrade function upgrade350. Let me know if this is not the case.

TimZ99 commented on src/templates/default/module/server/server/list.tpl.html:
```diff
@@ -8,6 +8,7 @@
 				<th scope="col">{{ label_label }}</th>
 				<th scope="col">{{ label_domain }}</th>
 				<th scope="col">{{ label_port }}</th>
+				<th scope="col">{{ label_protocol }}</th>
```
Could you update to the latest commit 🙈

coronas2k committed:
merged upstream/develop into develop

resolved conflicts in list.tpl.html by merging both changes.

coronas2k committed:
Added upgrader & changed version

Upgrades the install to include a protocol field.
Defined the upgrade version as 3.5.0

coronas2k commented:
I noticed a layout error after the recent changes, I'll submit a commit to update this.

coronas2k committed:
Fixed the protocol column width

coronas2k commented:
676e80d42f100353509178eb040e8b1d12df3b77 introduced inline styles which can be better handled with Bootstrap to be responsive & I noticed mismatching HTML tags but that's an issue for another day & another pull request. This will do in the meantime. 😹

TimZ99 commented on src/includes/psmconfig.inc.php:
```diff
@@ -30,7 +30,7 @@
  * Current PSM version
  */
-define('PSM_VERSION', '3.4.5');
+define('PSM_VERSION', '3.5.0');
```
Use `3.5.0.beta1` to prevent upgrade problems on next release. (only here)

TimZ99 commented on src/templates/default/module/server/server/view.tpl.html:
```diff
@@ -37,6 +37,12 @@
 						<dd class="col-md-8">{{ port }}</dd>
 					</dl>
 				</li>
+				<li class="list-group-item">
```
Moving this under `type` looks beter I think.

TimZ99 commented on src/lang/en_US.lang.php:
```diff
@@ -182,6 +182,7 @@ class="btn btn-primary">Next step</button></a> <br>This will open a chat
         'fieldset_permissions' => 'Permissions',
         'permissions' => 'Server will be visible for the following users',
         'port' => 'Port',
+        'protocol' => 'Protocol',
```
I see the issues coming already; udp gives false positive. There needs to be a helper under the protocol field warning about false positives due to firewalls etc.

TimZ99 commented on src/templates/default/module/server/server/update.tpl.html:
```diff
@@ -45,6 +45,15 @@
+				</select>
+			</div>
 			<!-- Custom port -->
 			{{ macro.input_field("number", "port", "port types typeService", "port", label_custom_port, edit_value_port, null, "5") }}
```
Protocol needs to be under custom port. When choosing for a custom port, you get the port selector, the protocol selector and after that the custom port field. Looks weird.

TimZ99 reviewed (commented):
Correct me if I'm wrong, but the udp checks will never fail right as udp ports don't give a response right?

coronas2k commented:
> Correct me if I'm wrong, but the udp checks will never fail right as udp ports don't give a response right?

Touché, let me play around with this more, might need something like #102 for UDP services to actually write to the socket and interpret the response, at least for now this is a WIP to achieve that goal.

TimZ99 commented:
Something like #102 will be in v4, don't think that it will come to v3 anymore.

coronas2k commented:
I clearly underestimated the effort in pulling this off. I've scoured high and low to look for a way to do it, due to the stateless nature of UDP the following would work:
- using exec to call nmap. (requires root to scan UDP)
- using a raw PHP socket to write an empty UDP response and check the ICMP response. (requires root to create a raw socket)

I see no way around it except if someone wants to run their php-fpm or Apache w/ mod_php under root which is definitely a bad idea... 🤦‍♂️

Edit: Or a Port checking API service

TimZ99 commented:
So did I. Also searched for it, tried some options, but same result.

Rayne commented:
> I see no way around it except if someone wants to run their php-fpm or Apache w/ mod_php under root which is definitely a bad idea... man_facepalming

It would be possible to allow to run only specific commands as privileged users with the help of a binary with `setuid` flag (not recommended IMHO) or a proper `sudo` configuration.

1. Configure `sudo` to allow the FPM pool user to run one specific command, script or binary as privileged user, e.g. a `/…/phpservermon/udp-service` shell script or PHP program. If the file approach would be used, then this file shouldn't be writable by regular users for security reasons.

2. `shell_exec` (or friends) to call the now configured `sudo` command

Another solution would be to SSH into a server (maybe localhost) that only allows to run specific commands (e.g. the previously mentioned script) on login.

coronas2k commented:
> 1. Configure `sudo` to allow the FPM pool user to run one specific command, script or binary as privileged user, e.g. a `/…/phpservermon/udp-service` shell script or PHP program. If the file approach would be used, then this file shouldn't be writable by regular users for security reasons.

Probing more into this it looks like it's possible by using setcap. I was able to do a UDP scan without "sudoing" see https://secwiki.org/w/Running_nmap_as_an_unprivileged_user

I imagine most users aren't running this script on shared hosting, so we could implement this as an extra configuration step to enable UDP scanning, with the security disclaimer.
